core/store/src/trie/outgoing_metadata.rs

```diff
@@ -0,0 +1,451 @@
+    /// scheduler will not be able to grant the bandwidth neeed to send
+    /// the receipts in this group.
+    pub size: u64,
+    /// Total gas of receipts in this group.
+    pub gas: u128,
+}
+
+impl ReceiptGroup {
+    /// Create a new group which will contain one receipt with this size and gas.
+    pub fn new(receipt_size: ByteSize, receipt_gas: Gas) -> ReceiptGroup {
+        ReceiptGroup::V0(ReceiptGroupV0 { size: receipt_size.as_u64(), gas: receipt_gas.into() })
+    }
+
+    /// Total size of receipts in this group.
+    pub fn size(&self) -> u64 {
+        match self {
+            ReceiptGroup::V0(group) => group.size,
+        }
+    }
+
+    pub fn size_mut(&mut self) -> &mut u64 {
+        match self {
+            ReceiptGroup::V0(group) => &mut group.size,
+        }
+    }
+
+    /// Total gas of receipts in this group.
+    pub fn gas(&self) -> u128 {
+        match self {
+            ReceiptGroup::V0(group) => group.gas,
+        }
+    }
+
+    pub fn gas_mut(&mut self) -> &mut u128 {
+        match self {
+            ReceiptGroup::V0(group) => &mut group.gas,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        // There are no receipts with size equal to zero.
+        // Total size being zero means that there are no receipts in this group.
+        let empty = self.size() == 0;
+        if empty {
+            assert_eq!(self.gas(), 0, "Gas should be zero for an empty group.");
+        }
+        empty
+    }
+}
+
+/// Parameters which control size and gas of receipt groups stored in the ReceiptGroupsQueue.
+/// There is a lower bound and upper bound on size and gas.
+/// By default receipts are added to the last group in the queue.
+/// A new group is started if:
+///     A) adding a new receipt would make the group exceed the upper bound
+///     B) the last group already has size or gas above the lower bound
+///
+/// This way we ensure that the size and gas of the groups are within the bounds,
+/// as long as size/gas of a single receipt doesn't exceed the upper bound.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ReceiptGroupsConfig {
+    /// All receipt groups aim to have a size above this threshold.
+    pub size_lower_bound: ByteSize,
+    /// All receipt groups aim to have a size below this threshold.
+    /// Should be no larger than `max_receipt_size`, otherwise the bandwidth
+    /// scheduler will not be able to grant the bandwidth neeed to send
+    /// the receipts in this group.
+    pub size_upper_bound: ByteSize,
+    /// All receipt groups aim to have gas above this threshold.
+    pub gas_lower_bound: Gas,
+    /// All receipt groups aim to have gas below this threshold.
+    pub gas_upper_bound: Gas,
+}
+
+impl ReceiptGroupsConfig {
+    pub fn default_config() -> Self {
+        // TODO(bandwidth_scheduler) - put in runtime config
+        ReceiptGroupsConfig {
+            size_lower_bound: ByteSize::kb(90),
```

Comment:
Should the lower bound be the same as the diff between two consecutive  requests? Is that 100kB?

Reply:
Diff between two consecutive values that can be requested is ~111kB
All groups will be at least as large as `lower_bound`, and that should be smaller than the diff between two values. I'm not sure what's the ideal value :thinking:  I guess the smaller we make the groups the better the granularity. I could reduce the group size to something like 50kB.